### PR TITLE
RegistryInterface: add locks to more methods

### PR DESCRIPTION
**Pull request by chrisbroome**
Several methods in the `RegistryInterface` class are reading and/or modifying member variables without locks. I've seen the `osqueryd --S` process terminate abnormally from this, and after debugging, the `RegistryInterface::remove` method was the culprit (more precisely `items_.erase(item_name)`).

This PR adds locks to the remaining methods in the class that read and/or modify member variables.

### CODE REVIEW

**Comment by chrisbroome** [comment]
Thanks for the quick feedback!

That's a good point and it looks plausible. Just trying to reason through the code, we're building a list of items to remove using a `ReadLock`, releasing that lock, removing external plugins, and then using a `WriteLock` to update the member collections.

I'm thinking the entire method could just have a single `WriteLock` at the top of it without the nested scoped ReadLock and WriteLock.

However, I must say, I'm no expert at multi-threaded c++, and I'm still very vague on osquery's internal runnable/runners.

**Comment by chrisbroome** [comment]
Also, I'm not sure of a great way to unit test the changes 🤷‍♂

**Comment by zwass** [comment]
Yes, my thought is also that a single WriteLock might be more appropriate. I don't have any suggestions for a quick test to these changes.

Usually @theopolis is the one to look to for guidance on thorny issues like this.

**Comment by chrisbroome** [comment]
Implemented changes for `RegistryInterface::removeExternal`

File: osquery/registry/registry_interface.cpp [hunk 1]
```diff
@@ -13,6 +13,8 @@
 
 namespace osquery {
 void RegistryInterface::remove(const std::string& item_name) {
+  WriteLock lock(mutex_);
```
Comment (theopolis) [inline osquery/registry/registry_interface.cpp]: You are holding a `ReadLock` in setUp and then calling `::remove`, which will try to grab a secondary `WriteLock` and deadlock.

File: osquery/registry/registry_interface.cpp [hunk 2]
```diff
@@ -268,29 +270,24 @@ Status RegistryInterface::addExternal(const RouteUUID& uuid,
 
 /// Remove all the routes for a given uuid.
 void RegistryInterface::removeExternal(const RouteUUID& uuid) {
+  WriteLock lock(mutex_);
```
Comment (theopolis) [inline osquery/registry/registry_interface.cpp]: Let's take a look at `registry_factory.cpp` where this is called and see if there are other locks that need to be synchronized.

Comment (theopolis) [reply osquery/registry/registry_interface.cpp]: - Look at calls into `::removeExternal`
- Look for the implementation of `::removeExternalPlugin`

File: osquery/registry/registry_interface.cpp (hunk 1 above)
Comment (theopolis) [reply osquery/registry/registry_interface.cpp]: One way of handling this is to create a `::removeUnsafe` method with the critical section, then provide the `::remove` public method that grabs the lock and waterfalls into `::removeUnsafe`. We'd have to look at the callsites of `::remove` (outside of this implementation) and see if that is OK.

Comment (chrisbroome) [reply osquery/registry/registry_interface.cpp]: Yeah that makes sense. Implemented a private `RegistryInterface::removeUnsafe` method with those changes.

**Comment by theopolis** [comment]
Thanks for the follow up Chris! I will take a look soon.

**Pull request merged**